File: frontend/src/lib/documents.ts
/**
 * Validação e formatação de documentos brasileiros — espelha
 * backend/app/core/documents.py.
 *
 * A validação aqui é conveniência de UX: avisa o advogado antes de gastar uma
 * request. A validação que vale é sempre a do servidor (CLAUDE.md, seção 16).
 *
 * O backend guarda e devolve **apenas dígitos**; a máscara é responsabilidade
 * da interface — por isso `formatCpf`/`formatCnpj`/`formatZipCode` existem só
 * para exibição, e o que se envia é sempre `stripNonDigits(...)`.
 */

export const CPF_LENGTH = 11;
export const CNPJ_LENGTH = 14;
export const ZIP_CODE_LENGTH = 8;

const CNPJ_FIRST_WEIGHTS = [5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2];
const CNPJ_SECOND_WEIGHTS = [6, ...CNPJ_FIRST_WEIGHTS];

export function stripNonDigits(value: string): string {
  return value.replace(/\D/g, "");
}

/** Dígito verificador pelo módulo 11, usado por CPF e CNPJ. */
function checkDigit(digits: string, weights: number[]): number {
  const total = weights.reduce((sum, weight, index) => sum + Number(digits[index]) * weight, 0);
  const remainder = total % 11;
  return remainder < 2 ? 0 : 11 - remainder;
}

export function isValidCpf(value: string): boolean {
  const digits = stripNonDigits(value);
  // Sequências repetidas ("111.111.111-11") passam no módulo 11 mas não são CPFs.
  if (digits.length !== CPF_LENGTH || new Set(digits).size === 1) return false;

  const first = checkDigit(digits.slice(0, 9), [10, 9, 8, 7, 6, 5, 4, 3, 2]);
  const second = checkDigit(digits.slice(0, 10), [11, 10, 9, 8, 7, 6, 5, 4, 3, 2]);
  return digits.slice(9) === `${first}${second}`;
}

export function isValidCnpj(value: string): boolean {
  const digits = stripNonDigits(value);
  if (digits.length !== CNPJ_LENGTH || new Set(digits).size === 1) return false;

  const first = checkDigit(digits.slice(0, 12), CNPJ_FIRST_WEIGHTS);
  const second = checkDigit(digits.slice(0, 13), CNPJ_SECOND_WEIGHTS);
  return digits.slice(12) === `${first}${second}`;
}

export function isValidZipCode(value: string): boolean {
  return stripNonDigits(value).length === ZIP_CODE_LENGTH;
}

export function formatCpf(value: string): string {
  const digits = stripNonDigits(value).slice(0, CPF_LENGTH);
  if (digits.length !== CPF_LENGTH) return digits;
  return `${digits.slice(0, 3)}.${digits.slice(3, 6)}.${digits.slice(6, 9)}-${digits.slice(9)}`;
}

export function formatCnpj(value: string): string {
  const digits = stripNonDigits(value).slice(0, CNPJ_LENGTH);
  if (digits.length !== CNPJ_LENGTH) return digits;
  return `${digits.slice(0, 2)}.${digits.slice(2, 5)}.${digits.slice(5, 8)}/${digits.slice(
    8,
    12,
  )}-${digits.slice(12)}`;
}

/** Formata CPF ou CNPJ conforme o comprimento — para exibir o que veio do backend. */
export function formatDocument(value: string | null | undefined): string {
  if (!value) return "—";
  const digits = stripNonDigits(value);
  if (digits.length === CPF_LENGTH) return formatCpf(digits);
  if (digits.length === CNPJ_LENGTH) return formatCnpj(digits);
  return value;
}

export function formatZipCode(value: string | null | undefined): string {
  if (!value) return "—";
  const digits = stripNonDigits(value);
  if (digits.length !== ZIP_CODE_LENGTH) return value;
  return `${digits.slice(0, 5)}-${digits.slice(5)}`;
}

/** Siglas de UF aceitas — espelha _BRAZILIAN_STATES do schema do backend. */
export const BRAZILIAN_STATES = [
  "AC", "AL", "AP", "AM", "BA", "CE", "DF", "ES", "GO", "MA", "MT", "MS", "MG",
  "PA", "PB", "PR", "PE", "PI", "RJ", "RN", "RS", "RO", "RR", "SC", "SP", "SE", "TO",
] as const;

export function isValidState(value: string): boolean {
  return (BRAZILIAN_STATES as readonly string[]).includes(value.trim().toUpperCase());
}

File: backend/app/core/documents.py
"""Validação e normalização de documentos brasileiros (CPF, CNPJ, CEP).

Toda escrita no banco guarda **apenas os dígitos**: o advogado digita
"529.982.247-25" ou "52998224725" conforme o humor, e sem normalizar na
entrada a checagem de duplicidade por documento
(uq_clients_tenant_id_document_number) deixaria os dois passarem como clientes
diferentes. A formatação com máscara é responsabilidade da interface
(frontend/src/lib/documents.ts).

Segurança: estas funções apenas validam formato e dígito verificador. Um CPF
estruturalmente válido não é prova de identidade e **nunca** deve ser tratado
como credencial de acesso — ver a docstring de app/models/client.py.
"""

import re

_NON_DIGITS = re.compile(r"\D")

CPF_LENGTH = 11
CNPJ_LENGTH = 14
ZIP_CODE_LENGTH = 8

#: Pesos do primeiro e do segundo dígito verificador do CNPJ. O CPF usa uma
#: progressão simples (10..2 e 11..2), calculada direto na função.
_CNPJ_FIRST_WEIGHTS = (5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2)
_CNPJ_SECOND_WEIGHTS = (6, *_CNPJ_FIRST_WEIGHTS)


def strip_non_digits(value: str) -> str:
    """Remove tudo que não for dígito de um texto.

    Args:
        value: Texto possivelmente formatado com máscara.

    Returns:
        Somente os dígitos, na ordem original.
    """
    return _NON_DIGITS.sub("", value)


def _check_digit(digits: str, weights: tuple[int, ...]) -> int:
    """Calcula um dígito verificador pelo módulo 11 usado em CPF e CNPJ.

    Args:
        digits: Dígitos base sobre os quais o verificador é calculado.
        weights: Pesos aplicados a cada dígito, na mesma ordem.

    Returns:
        O dígito verificador (0 quando o resto do módulo 11 é menor que 2).
    """
    total = sum(int(digit) * weight for digit, weight in zip(digits, weights))
    remainder = total % 11
    return 0 if remainder < 2 else 11 - remainder


def is_valid_cpf(value: str) -> bool:
    """Verifica se um CPF é estruturalmente válido (11 dígitos + verificadores).

    Args:
        value: CPF com ou sem máscara.

    Returns:
        True se o CPF tiver 11 dígitos, não for uma sequência de dígitos iguais
        e os dois dígitos verificadores conferirem.
    """
    digits = strip_non_digits(value)
    if len(digits) != CPF_LENGTH or len(set(digits)) == 1:
        # "111.111.111-11" e afins passam no módulo 11 mas não são CPFs reais.
        return False

    first = _check_digit(digits[:9], tuple(range(10, 1, -1)))
    second = _check_digit(digits[:10], tuple(range(11, 1, -1)))
    return digits[9:] == f"{first}{second}"


def is_valid_cnpj(value: str) -> bool:
    """Verifica se um CNPJ é estruturalmente válido (14 dígitos + verificadores).

    Args:
        value: CNPJ com ou sem máscara.

    Returns:
        True se o CNPJ tiver 14 dígitos, não for uma sequência de dígitos
        iguais e os dois dígitos verificadores conferirem.
    """
    digits = strip_non_digits(value)
    if len(digits) != CNPJ_LENGTH or len(set(digits)) == 1:
        return False

    first = _check_digit(digits[:12], _CNPJ_FIRST_WEIGHTS)
    second = _check_digit(digits[:13], _CNPJ_SECOND_WEIGHTS)
    return digits[12:] == f"{first}{second}"


def normalize_zip_code(value: str) -> str | None:
    """Normaliza um CEP para 8 dígitos.

    Args:
        value: CEP com ou sem máscara (ex.: "01310-100").

    Returns:
        Os 8 dígitos do CEP, ou None se o valor não tiver exatamente 8 dígitos.
    """
    digits = strip_non_digits(value)
    return digits if len(digits) == ZIP_CODE_LENGTH else None
